graveler/committed/diff.go

```diff
@@ -0,0 +1,207 @@
+package committed
+
+import (
+	"bytes"
+
+	"github.com/treeverse/lakefs/graveler"
+)
+
+type IteratorValue struct {
+	record *graveler.ValueRecord
+	rng    *Range
+}
+
+func (iv *IteratorValue) asDiff(typ graveler.DiffType) *graveler.Diff {
+	return &graveler.Diff{
+		Type:  typ,
+		Key:   iv.record.Key,
+		Value: iv.record.Value,
+	}
+}
+
+type diffIterator struct {
+	left       Iterator
+	right      Iterator
+	leftVal    *IteratorValue
+	rightVal   *IteratorValue
+	currentVal *graveler.Diff
+	err        error
+	state      diffIteratorState
+}
+
+type diffIteratorState int
+
+const (
+	diffItBeforeInit diffIteratorState = iota
+	diffItInitialized
+	diffItClosed
+)
+
+type diffIteratorCompareResult int
+
+const (
+	diffItCompareResultDone diffIteratorCompareResult = iota
+	diffItCompareResultSameRanges
+	diffItCompareResultSameIdentities
+	diffItCompareResultSameKeys
+	diffItCompareResultNeedStartRangeBoth
+	diffItCompareResultNeedStartRangeLeft
+	diffItCompareResultNeedStartRangeRight
+	diffItCompareResultLeftBeforeRight
+	diffItCompareResultRightBeforeLeft
+)
+
+func NewDiffIterator(left Iterator, right Iterator) graveler.DiffIterator {
+	return &diffIterator{
+		left:     left,
+		right:    right,
+		leftVal:  &IteratorValue{},
+		rightVal: &IteratorValue{},
+		state:    diffItBeforeInit,
+	}
+}
+
+func (d *diffIterator) diffIteratorNextValue(it Iterator) (*graveler.ValueRecord, *Range, error) {
```

Comment:
thought that next and next part will be func not a method and just return the err and in the code that call these functions we will assign it to d.err and return in case of an error.